### DISCUSSION TITLE
Рассинхронизация файла SavedTracks.json и getSavedTracks

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#249] [Помощь с ошибкой]
Обратил внимание, что у меня в массивах с лайками, стали часто попадаться треки без лайков. Стал изучать этот момент и оказалось, что в файле SavedTracks.json сотни не лайкнутых треков. Причем это не просто нелайкнутнатые версии лайкнутых треков, что иногда случается. А действительно, довольно большое количество треков, которых нет в Любимых. 
Более того. В самом файле меньше 3 тысяч треков, в то время когда лайков больше 13 тысяч. 

В принципе, интересна причина этого. 
У есть [скрипты](https://github.com/Chimildic/goofy/discussions/127) снимающие лайки с альбомов и с плейлистов, если они полностью прослушаны. Но лайки с самих треков они не снимают. Есть скрипт добавляющий треки в SavedTracks.json, но только после того как трек лайкнут. На всякий случай я этот момент из скрипта исключил. Больше  вроде ни какого автоматического вмешательства ни в файл SavedTracks.json, ни в сами лайки нет. Тем более какого-то удаления из файла. 

Я не знаю алгоритм формирования этого файла, но подозреваю, что это должна быть полная копия SavedTracks, с ежедневным обновлением. Т.е. новые лайки добавляются, разлайканые треки удаляются. Судя по всему, у меня эта синхронизация сломалась. 

Попробовал решить вопрос самостоятельно. Но не могу получить  массив getSavedTracks. Подозреваю из-за его размера. При получении например в пределах тысячи треков, все работает отлично. При большем, или максимальном значении появляется информация об ошибке 
```
Номер: 500
Адрес: https://api.spotify.com/v1/me/tracks?offset=100&limit=50&locale=RU
```
После этого каждые 2 секунды идет цикл таких сообщений с изменением значения положения трека offset на следующие 50 треков.
Естественно при достижении лимита времени, скрипт полностью останавливается не досчитав до конца. 

![2023-10-06_13-19-58](https://github.com/Chimildic/goofy/assets/76230194/399e428d-1060-48b0-91ac-ab776d016859)

UPD Все-таки за 300 секунд отработал скрипт с getSavedTracks. По информации из консоли offset=13150, т.е. все лайки. Но количество треков в полученном массиве почему-то всего 3750.  

Ну и собственно вопрос - что теперь делать? Каким-то образом получить список лайков, перезалить его в SavedTracks.json? Будет ли при этом соблюдаться история добавления лайков? Будут ли в дальнейшем синхронизировано содержание SavedTracks.json и getSavedTracks?

**Comment by Chimildic** [comment]
Лайки в файле обновляются каждый день единожды. Если новых лайков за день меньше 50, они просто добавляются к прошлым. Когда новых 50 и больше или сегодня понедельник, файл полностью перезаписывается данными от `Source.getSavedTracks`. 

Если ошибки 500 и паузы в 2 секунды укладываются в лимит выполнения, то в файл запишется не полный объем лайков. То есть запросы отправлены для всех, но успешные ответы только для части. goofy построен так, чтобы игнорировать некоторые ошибки дабы не прерывать скрипты на полпути. Отсюда паузы после ошибок. Иначе нестабильность ответов от Spotify ломает конечную цель - обновление плейлистов.

Учитывая что в файле остались треки, на которых уже нет лайка, файл давненько не перезаписывался. Или `Source.getSavedTracks` отдает устаревшие данные. Что маловероятно, но не удивился бы. Уже бывало, что в ответах приходит некорректное значение количества треков.

Какую дату покажет этот код? Наверное свежую. Иначе каждый запуск `runTasks_` длился бы долго или завершался ошибкой времени выполнения. Плюс надо посмотреть что в логах у  `runTasks_` в 01:00. Особенно прошлый понедельник.
```
function caseSyncSavedTracks() {
  let clerkProperty = JSON.parse(UserProperties.getProperties().ClerkTasks)
  console.log('Last success updateSavedTracks:', clerkProperty.updateSavedTracks)
}
```

Ошибки же 500+ это внутренние проблемы серверов Spotify. С ними ничего не сделаешь. Как писали на их [форуме](https://community.spotify.com/t5/Spotify-for-Developers/Being-returned-a-500-Server-error-from-v1-me-tracks-endpoint/m-p/5612898) этим летом, подобное может быть связано с большим количеством лайков. Например, могли потерять лимиты. Ведь в плейлист разрешают 10-11 тысяч, а лайки якобы безлимит или какие там правила сегодня, не знаю.

Теорию с лимитами относительно легко проверить. Руками унести пару тысяч нынешних лайков в обычный плейлист, чтобы не потерять. Снять с них лайки `Library.deleteFavoriteTracks` и вызвать `Source.getSavedTracks` на тест. Взяв за ориентир где нибудь 9-9.5 тысяч оставшихся лайков.

Поэтому поразбираться с `runTasks_` можно, но если `Source.getSavedTracks` не дает треков то и синхронизация поломается.

**Reply by Bedmit** [reply]
Код показал` Last success updateSavedTracks: 2023-10-01T22:32:25.704Z` В логах `runTasks_` ведет себя стандартно.
В 1:14 2.10.23 (этот понедельник) скрипт не отработал. Те самые ошибки каждые 2 секунды и в итоге "время ожидания истекло".
Добраться в логе до предыдущего понедельника уже не могу. Или не знаю как. 

![2023-10-07_11-29-31](https://github.com/Chimildic/goofy/assets/76230194/4385895d-c958-4a04-854f-8440494e7cbf)

**Comment by Bedmit** [comment]
Скрипя сердцем проверил теорию о превышении лимита в 10 т. треков. Потому что хоть лайки и можно потом восстановить, такой ценный ресурс как время их добавления, будет утерян. Но выхода нет, вопрос надо как-то решать.

Сперва удалил последние 3,5 т. треков, естественно сохранив их в файл.
```
function DeleteSavedTracks() {
let Saved = Source.getSavedTracks(3500);
  Cache.write('SavedBackup.json', Saved);
  Library.deleteFavoriteTracks(Saved);
}
```
После этого удалось через `Source.getSavedTracks` получить все 9,5 т. оставшихся лайков. 
На сколько я понимаю, результат говорит о том, что со стороны Spotify ограничений в лайках и отдаче их через API нет. Проблема исключительно со стороны Google Apps Script, с его ограничением обработки одного скрипта на 6 минут. Он просто не успевает посчитать такое количество треков. Не знаю, возможно сам скрипт можно как-то оптимизировать, например делать паузу в 1 секунду. Но все равно когда-то можно упереться в потолок. Если на глазок, то это 50 * 1 * 60 * 6 = 18 т. треков.

Теперь соответственно вопросы:
1. Как принудительно восстановить файл SavedTracks.json?
2. Как потом добавить в него же удаленные 3,5 т. лайков?
3. Что теперь в принципе делать с синхронизацией, если я верну лайки обратно?

**Reply by Bedmit** [reply]
Дополнительно к выше предложенному решению. Если версия, что `Source.getSavedTracks` стабильно может отработать в одном скрипте не больше 10 т. треков верна, то может есть возможность разделить эту задачу на несколько скриптов? Сперва получить и обработать первые 10 т. треков, а если их больше следующим скриптом вторые 10 т. треков и т.д. Первое я понимаю реализуется просто `Source.getSavedTracks(10000)`. Дальше видимо через позицию трека в плейлисте `offset`. Заранее извиняюсь, если звучит слишком по дилетантски.

**Reply by Chimildic** [reply]
Результат другой. Spotify разрешает добавлять больше 10 тысяч, но по API сыпет ошибки. То есть было 13к и все треки не давало, стало 9.5к все треки отдало. Причем интересно, что при 13к ошибки начинаются с 6650 судя по скриншоту, а при 9.5к то же место проходит без ошибок. Прям явно выраженная проблема с лимитами с их стороны.

Уменьшение паузы не поможет. Лайки запрашиваются пачками. Сразу с десяток запросов за раз (параллельно). Поэтому без ошибок лайки приходят быстро. То что мы видим в логах паузу в 2 секунды это единичная попытка повторить запрос вне пачки. Когда пришли ответы на всю пачку, goofy читает их и ошибки отправляет на повтор. Часто помогает от рандомных ошибок. Отсюда получаем: много ошибок = много единичных попыток = долгое время выполнения. 

Без пауз получили бы лишь часть. Как пример из основного поста - 3750 треков успевшие за 300 секунд. С паузами чаще не успевает, но итог все равно был бы 3750. В обоих случаях это не то что нужно. При рандомных ошибках (вне контекста с лайками) повтор наоборот помогает получить данные. Отсюда вывод, что все таки лучше с паузами.

По вопросам, тут что не предложи костыль на костыле будет. Поломают идею кэша или треки не будут актуальными. Как следствие ломая остальные алгоритмы, где лайки важны. 

- Например, отключить синхронизацию в runTasks_ закомментировав Clerk-задачи и руками соединить 9.5к с 3.5к в один кэш файл с тем же именем. Получается треки будут устаревать, но хотя бы не перезаписываться. И остальной код менять не придется, где читается этот файл. Ноль синхронизации, заморозка, но все 13к в обороте.
- Другой вариант, ограничить максимальный объем, который просим при перезаписи кэш файла. Меньше лайков в обороте, но хотя бы актуальные. При этом удалить файл с диска, чтоб в этот же день сработала задача на создание файла. Возникает проблема какое число выбрать? Вон при 13к для теста 3.5к спокойно отдали, а когда все просишь на 6.5к ошибки начинаются ¯\_(ツ)_/¯

```js
// код внутри runTasks_
function updateSavedTracks(tracks) {
    tracks = tracks || Source.getSavedTracks(); // сюда вписать число
    Cache.compressTracks(tracks);
    Cache.write('SavedTracks.json', tracks);
}
```

Учитывая что проблему заметили только сейчас, можно надеется что это временно. Возможно не лишним будет им на форум написать и натравить голосов из телеграм чата и 4pda чтоб заметили проблему с большим количеством лайков.

**Reply by Chimildic** [reply]
> Дополнительно к выше предложенному решению. 

Для их стороны нет разницы какой скрипт отправляет запрос. Что подряд проси, что дели на части. В обоих случаях конечные аргументы одинаковые.

Как и написал выше, сомневаюсь, что при 13к функция `Source.getSavedTracks(10000)` сработает. Потому что сыпется уже на 6к.

**Reply by Bedmit** [reply]
Тут требуется маленькое уточнение. Чтобы возможно мой ни совсем корректный отчет, не повлек за собой неверные выводы.

Ошибки начинаются не с 6650 трека. Скриншот сделан произвольно внутри всей ленты ошибок. На самом деле они начинаются прямо с 50 трека. 
![2023-10-07_15-49-37](https://github.com/Chimildic/goofy/assets/76230194/97cbdda9-52a4-45e6-acbd-01dc8d042f32)

Причем ошибки есть всегда и сразу, вне зависимости меньше 10 т. треков в лайках, или больше. Они могут пропадать при принудительном ограничении получаемого массива. Например `Source.getSavedTracks(600)` не давал ошибок и при 13 т. треков в лайках и при 9,5. А при `Source.getSavedTracks(1000)` они уже могут появиться, но с каким-то бОльшим шагом ( offset=300 , потом offset=850 и т.д.). Хотя потом все равно цикл уходит в шаг в 50 треков. 
Т.е. получается отсутствие ошибок и способность получить треки пачкой, зависит не только от общего количества лайков, а и от того количества которое требуется получить. Хотя возможно при бОльшем общем количестве лайков, размер пачки которую скрипт обработает без ошибок и не уйдет сразу в цикл, будет гораздо меньше.

При попытке получить все лайки, если их количество довольно большое (какое?) процесс в любом случае начинается с ошибки и уходит в цикл offset=50. Так было и при 13 т., так же осталось и при 9,5 т. Вот только при 9,5 т. весь процесс поместился в 6 минут, а при 13 т. нет. 
Но даже когда он поместился, не понятно, почему при успешном завершении и при отображении в информации об ошибках, что offset=13150 т.е. подсчет дошел до последнего трека, на выходе осталось только 3750 треков. Я вообще не очень понимаю, почему если в информации об ошибке написано, что обработка дошла до позиции такого-то трека, это же количество треков не передается в массив? А тем более когда скрипт полностью успешно отработал.

Сейчас провел еще эксперимент. Вернул обратно все лайки и попытался получить максимальное количество треков через `Source.getSavedTracks(X)`. При 3000 получил 2800 треков, при 5000 - 4600, при 6000 - 5650, при 7000 скрипт не отработал (дошел до 6400 треков), при 8000 - 1300, при 10000 скрипт не отработал. Самое интересное, что при повторении вышли те же цифры.


Отсюда вопрос. Раз какое-то количество треков все равно можно получить, не смотря на информацию об ошибках, есть ли возможность получать из `Source.getSavedTracks` треки кусками (например с 1 по 5000, потом с 5000 по 10000) в разных скриптах, чтобы не упереться в лимит времени, и соединять их в одном файле?

UPD Кстати сейчас опять убрал 3,5 т. лайков и запустил на оставшихся 9,6 т. лайков `Source.getSavedTracks()
` Скрипт не отработал.

**Reply by Bedmit** [reply]
На счет решения на данный момент. Хотя бы и временного, и с костылями. 

Первое точно не устраивает. Скрипты как раз и построенные на использовании добавляющихся лайков перестанут работать.

А со вторым решением, правильно я понимаю, что если я выберу например 5000 треков (как показали изыскания выше, более-менее оптимальный размер с запасом), то ежедневно будет добавляться количество лайков добавленных в течении дня (в пределах 50), а раз в неделю перепроверяться последние 5000 лайков? Все более ранние лайки так и останутся в SavedTracks.json, но перепроверяться на предмет снятия лайка, уже не будут. Но тогда ведь мне надо как-то засунуть все 13 т. лайков в файл, в первый раз? 
Или я не правильно понимаю суть решения?

Что-то мне подсказывает, что в SavedTracks.json всегда будет 5000 последних лайков. Что тоже категорически не устаивает. А нельзя ли подредактировать код, чтобы и последние 5000 лайков проверялись и все остальные лайки тоже оставались в  SavedTracks.json? Т.е. вместо `Cache.write('SavedTracks.json', tracks)` использовать какой-то аналог функции saveWithUpdate. То что проверили - изменили, а то что не смогли проверить - так и оставили.

**Reply by Chimildic** [reply]
Маленькое уточнение сильно меняет расклад.

> Я вообще не очень понимаю, почему если в информации об ошибке написано, что обработка дошла до позиции такого-то трека, это же количество треков не передается в массив?

`getSavedTracks` отправляет один запрос на 50 треков. Если в аргументе функции нет значения или оно больше 50, из первого запроса берутся данные об общем количестве треков. На этой основе заранее создаются ссылки ввида `?offset=50&limit=50`. Массив ссылок делится на пачки по N-штук. Отправляем первую пачку и ждем ответы на каждый запрос из нее. Читая ответы нарываемся на ошибку. Делаем паузу и повторно отправляем запрос. Теперь не всю пачку, а только этот конкретный запрос. Если снова приходит ошибка, считаем ответ для этого запроса пустым. Потому что уже дважды пытались его получить. И продолжаем смотреть остальные ответы из пачки. 

Отсюда появляется разница между запрошено 3к получено 2.8к или запрошено 13к получено 3.7к. Все чего не хватает - дважды дало ошибку. Нет смысла стучать туда до бесконечности. Это в целом общее поведение почти для всех запросов. Чтобы дать шанс скриптам завершится как задумано, а не сидеть на одном месте из-за случайно вылетевшей ошибки. 

Таким образом, лог с ошибкой про `offset=13150` говорит что **ошибка получена в пачке**, а пауза говорит что будет попытка **единичного** повтора. Информации о том была ли повторная попытка успешной - нет в логах. В этом кроется ошибка _отчета_. Я подумал, что при 9.5к все прошло без ошибок. 

> Вот только при 9,5 т. весь процесс поместился в 6 минут, а при 13 т. нет.

Поэтому _поместился_ совсем не значит, что получил все.

Раз и при 13к и при 9.5к сыпятся ошибки со старта - теория с лимитами неверная. По крайней мере, из вне к этим лимитам нам не подстроиться. Для нас в целом "болеет" точка доступа к лайкам. 

> есть ли возможность получать из Source.getSavedTracks треки кусками

Мне кажется бессмысленно перекладывать с больной головы на здоровую. Ошибка прямо со старта. С разным шагом доходящим до "каждый первый" при любом количестве лайков. Тут целый госпиталь с костылями нужен, причем им самим.

Наверное могу лишь предложить создать новое приложение в [Spotify Dashboard](https://developer.spotify.com/dashboard), чтобы получить новые `client id` и `client secret` для конфиг файла goofy. Вдруг это как-то влияет на то какой сервер отвечает на запросы.
И репорт на их форум тоже подойдет.

**Reply by Chimildic** [reply]
> Что-то мне подсказывает, что в SavedTracks.json всегда будет 5000 последних лайков.

Да, файл перезаписывается целиком. Учитывая новую вводную, пока эти 5к не приходят без единой ошибки - менять что-то нет смысла.

**Reply by Bedmit** [reply]
Если создать новое приложение, это на данный момент единственно возможный способ попытаться как-то решить вопрос, то наверное придется. Правда не совсем понимаю как это сделать и потерей чего это может грозить.

**Reply by Bedmit** [reply]
А если количество треков будет например 500? Они считаются одной пачкой без ошибок, при любом размере лайков. Конкретно для моих скриптов это тоже вполне себе выход. Ибо снимаю лайки я довольно редко, а вот использую их массив очень часто.

**Reply by Chimildic** [reply]
> создать новое приложение

Потери никакой нет. Просто выдача прав по другому приложению. 
Они немного изменили форму создания приложения. Хотят ссылку сразу.

1. Этот код сбросит авторизацию и даст ссылку

```js
function logCallbackUrl() {
  Auth.reset()
  console.log(`https://script.google.com/macros/d/${encodeURIComponent(ScriptApp.getScriptId())}/usercallback`)
}
```

2. Нажимаем `create app` в [Dashboard](https://developer.spotify.com/dashboard), `name` и `description` произвольно, `redirect uri` из первого шага. 

3. После создания в settings будет `client id` и `client secret`. Их копируем в config вместо текущих и вызываем функцию `setProperties` для их сохранения.

4. В управлении развертываниями создаем новое и заходим по ссылке в веб приложение, чтобы подтвердить выдачу прав.

> если количество треков будет например 500?

Между 500 и 5000 почти нет разницы по коду. Отличается сколько пачек будет отправлено. Но и это не имеет смысла потому что выше был пример, когда первая же пачка на первые же 50 треков дало ошибку. 

Этим кодом можно поиграть выставляя разные смещения. Скорее всего будут ошибки как и при последовательном смещении.
```js
function caseSavedTracks() {
  let offsets = [50, 300, 400, 700, 900, 5000, 7400, 7450, 8000, 9400]
  let urls = offsets.map(value => `https://api.spotify.com/v1/me/tracks?limit=50&offset=${value}`)
  SpotifyRequest.getAll(urls)
}
```

За размер пачки отвечает `REQUESTS_IN_ROW` в конфиг файле. Можно уменьшить до 10-20 для теста (количество одновременных запросов). Это влияет на время выполнения. Чем меньше значение, тем дольше время.

**Reply by Bedmit** [reply]
Решил начать с простого - создать новое приложение. Ни каких исправлений это не принесло, увы. Попутно вопрос. Что лучше - удалить старое, или вернуться на него? Если да, то как?

По поводу 500 треков, может опять что-то написал не совсем корректно, но при **любом** количестве лайков, такой запрос **всегда** считается без ошибок, сразу одной пачкой. В принципе и большее количество считает все треки, хотя и могут выскочить сообщения об ошибке. Видимо часть треков все равно выдается при повторном запросе. 

Но для моих задач достаточно проверки и последних 500 треков. Поэтому хотел бы попробовать изменить runTasks_ примерно так
```cs
function updateSavedTracks(tracks) {
    tracks = tracks || Source.getSavedTracks(500);
    Cache.compressTracks(tracks);
    Cache.append('SavedTracks.json', tracks, 'begin');
    
    let tracksAll = Cache.read('SavedTracks.json');
    Filter.dedupTracks(tracksAll);
    Cache.write('SavedTracks.json', tracksAll);
}
```
Стоит попробовать такой вариант? 

Вариант с изменением смещения, каюсь, не понял.

**Reply by Chimildic** [reply]
> удалить старое, или вернуться на него

Оно места не занимает. И как выяснилось одинаковое. Поэтому нет разницы. Пусть останется, чтоб ничего не делать больше. 
Если хочется вернутся, то схема та же. Скинуть авторизацию, вернуть прошлые client_id с secret, сохранить и права выдать.

> Вариант с изменением смещения, каюсь, не понял.

Просто для примера позапускать функцию. По желанию, поменять или добавить чисел в массив. Чтобы проверить как реагирует, когда смещение не подряд +50, а с большим и разным шагом.
Как раз к вопросу о том, можно ли делать несколько функций с разными смещением. Если даст ошибок, то можем и не пытаться.

>  при любом количестве лайков, такой запрос всегда считается без ошибок

По умолчанию пачка равна 40 запросам (Значение REQUESTS_IN_ROW). 500 треков это 10 запросов по 50 треков (меньше 1 пачки). 5к это 100 запросов (2 с копейкой пачки). Может быть Spotify ломается из-за наплыва параллельных запросов? Когда 40 приходит падает в ошибки, а когда 10 отдает корректно. Похоже на зацепку. 

Попробуем уменьшить REQUESTS_IN_ROW до 10 в конфиг файле и запросить все лайки (не забывая запустить setProperties).

**Reply by Bedmit** [reply]
Таки да, получилось!!! Все 9695 треков до копейки, без единой ошибки. Причем довольно бодро посчитал.
Сейчас попробую вернуть лайки до 13 к и посчитать с ними.

**Reply by Bedmit** [reply]
Добавил. Посчитало все 13195 треков. По дороге выдавало сообщения об ошибках, но считало с довольно большими и неравномерными шагами. Тем не менее в результате ни одного трека не утеряно!
Не знаю, как будет при большем количестве лайков и вообще, как будет себя вести в дальнейшем. Но сейчас это отличный результат. 
Как всегда большое спасибо за поиск решение очередной проблемы!

Я полагаю, мне сейчас есть смысл запустить восстановление `SavedTracks.json`?
```
    tracks = Source.getSavedTracks();
    Cache.compressTracks(tracks);
    Cache.write('SavedTracks.json', tracks);
```

**Reply by Chimildic** [reply]
Единственное я не знаю насколько существенно это влияет на остальные скрипты с точки зрения времени выполнения. Например там, где собираются треки из больших плейлистов (внутри тот же принцип, пачками). В первых версиях goofy было последовательно друг за другом, а с появлением REQUESTS_IN_ROW резко ускорилось. И до сих пор это единственный способ параллельности. Google не хочет добавлять асинхронность в GAS.

Поэтому нужно понаблюдать за остальными скриптами, не выпадет ли в лимит выполнения. Плюс можно поиграть со значением. Увеличив например до 20. 

> запустить восстановление SavedTracks.json

Да, можно как в примере кода. Или просто удалить файл с диска, он этой ночью сам создастся из-за отсутствия файла.

**Reply by Bedmit** [reply]
Все восстановил. Даже все плейлисты пообновлял. Теперь сразу видно, что суть скриптов выполнена верно. Особых изменений при обработке скриптов по времени, не заметил. Буду наблюдать, появятся ли они в дальнейшем.
Еще раз спасибо!

На счет тикета на форуме по этой проблеме. Не особо верю в его эффективность если честно, но если надо на 4pda без проблем помогу с продвижением.

**Reply by Bedmit** [reply]
Дополнительно, в подтверждении работы автоматики. 

В этот понедельник `runTasks_`  отработал `updateSavedTracks` полностью. Информация об ошибках с неравномерным шагом есть, скрипт уложился почти в 2 минуты и посчитал все лайки. Хотя, думаю количество ошибок и время выполнения может быть разным.

Поэкспериментировал сейчас немного с REQUESTS_IN_ROW, со значениями 5, 10 и 20. Время выполнения:
5 - 77 сек
10 - 90 сек
20 - 109 сек (100 треков не досчиталось)
Информация об ошибках при любом значении.

В итоге решил оставить 10. За эти дни, влияния изменения этого параметра на работу других скриптов, не заметил.
